Remove CSS that causes conflict with theme.json

ndiego opened the pull request:
## Description
Now that theme.json supports margin for column blocks, the removed CSS takes precedent over any configured theme.json margin styles. This impedes theme.json adoption. Since this CSS served no real purpose anyway, I propose that it is removed.

## How has this been tested?
This has been tested with TT1 blocks, TT2 and custom FSE themes and Gutenberg 11.9.

## Types of changes
Bug fix

## Checklist:
- [X] My code is tested.
- [X] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [X] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [X] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [X] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [X] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [X] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

ndiego committed:
Remove CSS that causes conflict with theme.json

Now that theme.json supports margin for column blocks, the removed CSS takes precedent over any configured theme.json margin styles. This impedes theme.json adoption. Since this CSS served no real purpose anyway, I propose that it is removed.

jasmussen commented:
Thank you for the PR! Thanks for the ping!

So, here's what's up. For non-block themes, we have a blanket rule that outputs top and bottom margins for everything that's defined to be a _block_ — in the editor only. This rule:

```
html :where(.wp-block) {
	...

	// Provide every block with a default base margin. This margin provides a consistent spacing
	// between blocks in the editor.
	margin-top: $default-block-margin;
	margin-bottom: $default-block-margin;
```

The problem here is, what is defined as a "block" is pretty generic, and includes both the _columns_ block, and the _column_ block. And so if we remove the rule as this PR does, the following happens in TT1:
![tt1](https://user-images.githubusercontent.com/1204802/141428210-a4054ceb-74cf-46cf-a979-d85c76c0d09c.gif)

It's fine in a block theme, because there aren't those pre-existing margins applied in the editor:
![block theme](https://user-images.githubusercontent.com/1204802/141428245-5f777b42-5a6a-4228-9fd9-163b955cbc41.gif)

And so if we remove this rule, classic themes that use columns will look different in the editor. So I think we need a different fix: keep the rule in the editor.scss file, so that it applies only to the editor, but lower specificity. For example, this rule:

```
html :where(.wp-block-column) {
    margin-top: 0;
    margin-bottom: 0;
}
```

Without testing theme.json, this _should_ have low enough specificity that anything you add there should override it. I consider this `:where` trick to be a way to apply [soft defaults](https://codepen.io/joen/pen/abJVrGX). 

For me it works *:

![where](https://user-images.githubusercontent.com/1204802/141428738-73b888aa-c6e6-4e29-b0cc-c9003542d6b8.gif)

* It works if we ignore the blanket `[data-block]` margins that TT1 outputs, but those should probably be revisited in the first place. So the fix above should work for _most_ classit themes that don't provide blanket editor block margin styles like that, which feels like the right balance to find.

ndiego committed:
Add back styling but with lower specificity.

ndiego commented:
Ahh I see. Thanks for the additional information. I have added your recommended CSS and it works well with theme.json configured margins. I also tested in TT1 and removed the blanket `[data-block]` margins that it outputs and all still looks good. Should I put together a PR to remove that in TT1, or to you think there may be other reasons to keep it?

Anyway, this should be good to go unless you have any further concerns. Thanks for reviewing!

ndiego committed:
Fix linting errors.

jasmussen commented:
Thanks for the fix, kudos on the extra comment. Codewise this looks good and should work well. I don't have the bandwidth to test this one right now, but things to verify is to smoke test a few classic themes. A way to test would be a 2 columns block with paragraphs inside, then paragraphs before and after. This PR should look the same as trunk — with the notable exception of TT1 which has that rule I mentioned.

ndiego commented:
I have gone ahead and tested with Twenty Twenty-One, Twenty Twenty, and Twenty Nineteen as well as TT1 Blocks, and TT2 with custom margins in theme.json. All works as expected.

jasmussen commented:
I've restarted the tests as the failure seems unrelated. If someone else is able to give this one a quick sanity chec, the code should be good to go.

Mamaduka commented:
Hi, @ndiego

Can you rebase this branch on top of the trunk? Failing E2E is skipped for now, so this should fix the checks.

ndiego commented:
@Mamaduka just rebased. Let me know if this needs anything else and hopefully the checks all pass 🤞

ndiego commented:
@Mamaduka and @jasmussen I have retested this and it's still good to go. Anyway we can get this fix merged?

jasmussen commented:
Sorry for missing this. It's in bug territory so it can still land. I'll give this one a spin again tomorrow and see if I can't green light it, if no-one else beats me to it. Thanks for your work.

jasmussen reviewed (approved):
Thanks for the PR. I gave it a thorough test. 

To recap, the problem to solve is that in classic themes, a blanket top/bottom margin is output on every _block_, including _column_ (singular). Those rules are not output in pure block themes, but are in classic themes. In that light, this PR solves a problem for classic themes that use theme.json.

Block theme test first — and in this case there should be no difference to trunk:
<img width="402" alt="it takes in a block theme" src="https://user-images.githubusercontent.com/1204802/144575340-17fb4451-1de9-4259-86e1-95b43d85adf5.png">
<img width="739" alt="editor after, block theme" src="https://user-images.githubusercontent.com/1204802/144575342-88a3e81e-8c02-46cf-8299-73ac1153e575.png">
<img width="1058" alt="frontend after, block theme" src="https://user-images.githubusercontent.com/1204802/144575346-8310ad0d-0b2e-4539-ba3e-6d13990a3679.png">

All works, and the code takes even if it doesn't unset anything. The use of :where means low specificity.

Classic theme:
<img width="773" alt="editor after, classic theme" src="https://user-images.githubusercontent.com/1204802/144575465-3cdb0ee5-ad88-473a-bdaf-b5c80091120d.png">

<img width="797" alt="frontend after, classic theme" src="https://user-images.githubusercontent.com/1204802/144575483-53ae6934-db8b-4a2b-8425-0d26c784069d.png">

The rule does take effect:

<img width="2032" alt="it takes in a classic theme" src="https://user-images.githubusercontent.com/1204802/144575509-a11607e9-cab8-481d-981a-48cf9c4af97e.png">

Here's what it looks like without the rule:

<img width="878" alt="editor without the rule" src="https://user-images.githubusercontent.com/1204802/144575527-9fee8bf8-202e-4a4c-a73d-299372367e6e.png">

<img width="346" alt="editor without the rule code" src="https://user-images.githubusercontent.com/1204802/144575530-522a208b-720b-435f-ae32-9112c971cf9c.png">

Note that I had to rebase to test this PR. This is good to land. My only question is, should we add a small additional comment to note that this rule is meant for classic themes that use theme.json, and isn't needed (but is harmless) for block themes?

Pre-approving, but would love that small comment added, then you can land this. Thanks again!

noisysocks commented:
Merging this as it has the backport label and I see that it's approved.

ndiego commented:
Thanks @noisysocks, and thanks @jasmussen for the additional review. I tested again just now and all looks good.
